cql3/statements/modification_statement.cc

```diff
@@ -531,6 +532,20 @@ modification_statement::validate(service::storage_proxy&, const service::client_
     }
 }
 
+void modification_statement::validate_timestamp(const query_options& options) const {
+    if (attrs->is_timestamp_set()) {
+        static constexpr int64_t MAX_DIFFERENCE = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::days(3)).count();
+        auto now = std::chrono::duration_cast<std::chrono::microseconds>(db_clock::now().time_since_epoch()).count();
+
+        auto timestamp = get_timestamp(now, options);
+
+        if (timestamp - now > MAX_DIFFERENCE) {
+            throw exceptions::invalid_request_exception("Cannot provide a timestamp more than 3 days into the future. If this was not intended, "
+            "make sure the timestamp is in microseconds");
+        }
+    }
+}
+
```

Comment:
This looks like code duplication. I think you can move it to query_options.

Perhaps it can be called even earlier, from transport/server.cc, when the frame is parsed and the timestamp extracted.

Reply:
While query_options are indeed populated early, server.cc is not aware of the exact type of the statement and (at least to my knowledge) checking what was provided in the `USING TIMESTAMP` clause directly would require ugly checks and casting.

Now, as for the code duplication and moving this to query_options - apart from them, we also determine the timestamp based on the attributes of the statement and delegating that to query_options (which is used to populate the statement, not the other way around) would create a very counter-intuitive back and forth (pass the options to statement -> get timestamp using options -> pass that timestamp back to options for validation).
I can extract that code to a separate function, standalone in some namespace, or perhaps a static somewhere up the statement class hierarchy tree. But, is it really worth it? I'd argue that it's a small piece of code in only 2 places (I don't think `USING TIMESTAMP` will appear in many more statement types soon) and it provides a natural separation of concerns as it is right now.

Reply:
I had the same observation as @avikivity but couldn't convince myself one way or the other. Fundamentally this code does a well-separated check so it wouldn't hurt to wrap it into a well-named and well-commented free function.

Reply:
I would avoid doing it in a class hierarchy.

Reply:
My suggestion to move to transport or query_options is nonsensical, you don't have access to batch_statement::_attrs there.

But a free function to reduce duplication is a good idea. It's a small amount of code but a free function is not adding any complexity (and de-duplicates a constant so we don't have to change it twice if we ever want to).